### PR TITLE
Catch wrong identifier type

### LINKED ISSUE #3100: Catch wrong identifier type
Currently, when I have a UUID identifier type and an entity that has a foreign key attribute to this identifier, and I try to post an invalid UUID type I am getting error 500 and it throws **Doctrine\DBAL\Types\ConversionException**.

We need to catch this error and throws a custom exception implementing **ApiPlatform\Core\Exception\InvalidArgumentException**.
The code which throws the error is https://github.com/api-platform/core/blob/v2.4.7/src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php#L79-L84

Note: I am using plain identifier.

Stack trace:
[trace.txt](https://github.com/api-platform/api-platform/files/3659736/trace.txt)

More details we talked in the Slack api-platform channel: https://symfony-devs.slack.com/archives/C39FKU9AL/p1569504477289300?thread_ts=1569441831.241100&cid=C39FKU9AL

### PR DESCRIPTION
**Pull request by Zowac**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | fix #3100 
| License       | MIT

### CODE REVIEW

File: src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php [hunk 1]
```diff
@@ -76,11 +77,15 @@ private function normalizeIdentifiers($id, ObjectManager $manager, string $resou
...
-            if ($isOdm && null !== $doctrineTypeName && MongoDbType::hasType($doctrineTypeName)) {
-                $identifier = MongoDbType::getType($doctrineTypeName)->convertToPHPValue($identifier);
+            try {
+                if ($isOrm && null !== $doctrineTypeName && DBALType::hasType($doctrineTypeName)) {
+                    $identifier = DBALType::getType($doctrineTypeName)->convertToPHPValue($identifier, $platform);
+                }
+                if ($isOdm && null !== $doctrineTypeName && MongoDbType::hasType($doctrineTypeName)) {
+                    $identifier = MongoDbType::getType($doctrineTypeName)->convertToPHPValue($identifier);
+                }
+            } catch (\Exception $e) {
```
Comment (alanpoulain) [inline src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php]: The catch is too broad.

File: src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php [hunk 2]
```diff
@@ -76,11 +77,15 @@ private function normalizeIdentifiers($id, ObjectManager $manager, string $resou
...
+                throw new InvalidArgumentException('Invalid argument, it must be an valid UUID');
```
Comment (alanpoulain) [inline src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php]: The message is too specific. Also, add the previous exception.

**Comment by alanpoulain** [comment]
Could you add a unit test?

File: src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php [hunk 3]
```diff
@@ -13,12 +13,14 @@
...
 namespace ApiPlatform\Core\Bridge\Doctrine\Common\Util;
 
+use ApiPlatform\Core\Exception\InvalidArgumentException;
 use ApiPlatform\Core\Exception\PropertyNotFoundException;
 use Doctrine\Common\Persistence\ObjectManager;
 use Doctrine\DBAL\Types\Type as DBALType;
 use Doctrine\ODM\MongoDB\DocumentManager;
 use Doctrine\ODM\MongoDB\Types\Type as MongoDbType;
 use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\DBAL\Types\ConversionException;
```
Comment (teohhanhui) [inline src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php]: CS: Sort alphabetically.

File: src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php [hunk 4]
```diff
@@ -76,11 +78,15 @@ private function normalizeIdentifiers($id, ObjectManager $manager, string $resou
...
-                $identifier = MongoDbType::getType($doctrineTypeName)->convertToPHPValue($identifier);
+            try {
+                if ($isOrm && null !== $doctrineTypeName && DBALType::hasType($doctrineTypeName)) {
+                    $identifier = DBALType::getType($doctrineTypeName)->convertToPHPValue($identifier, $platform);
+                }
+                if ($isOdm && null !== $doctrineTypeName && MongoDbType::hasType($doctrineTypeName)) {
+                    $identifier = MongoDbType::getType($doctrineTypeName)->convertToPHPValue($identifier);
+                }
+            } catch (ConversionException $e) {
+                throw new InvalidArgumentException(sprintf('Invalid identifier "%s" value provided.', $propertyName), $e->getCode(), $e);
```
Comment (teohhanhui) [inline src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php]:
```suggestion
                throw new InvalidArgumentException(sprintf('Invalid value "%s" provided for an identifier.', $propertyName), $e->getCode(), $e);
```

File: src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php (hunk 3 above)
Comment (Zowac) [reply src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php]: Done :)

File: src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php (hunk 4 above)
Comment (Zowac) [reply src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php]: Done, thank you :)

File: src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php [hunk 5]
```diff
@@ -76,11 +78,15 @@ private function normalizeIdentifiers($id, ObjectManager $manager, string $resou
...
-                $identifier = MongoDbType::getType($doctrineTypeName)->convertToPHPValue($identifier);
+            try {
+                if ($isOrm && null !== $doctrineTypeName && DBALType::hasType($doctrineTypeName)) {
+                    $identifier = DBALType::getType($doctrineTypeName)->convertToPHPValue($identifier, $platform);
+                }
+                if ($isOdm && null !== $doctrineTypeName && MongoDbType::hasType($doctrineTypeName)) {
+                    $identifier = MongoDbType::getType($doctrineTypeName)->convertToPHPValue($identifier);
+                }
+            } catch (ConversionException $e) {
+                throw new InvalidArgumentException(sprintf('Invalid value "%s" provided for an identifier.', $propertyName), $e->getCode(), $e);
```
Comment (soyuka) [inline src/Bridge/Doctrine/Common/Util/IdentifierManagerTrait.php]: https://github.com/api-platform/core/blob/master/src/Exception/InvalidIdentifierException.php ? :D

**Comment by soyuka** [comment]
> Could you add a unit test?

I'm not sure it's possible to mock the static DBALType to make it throw right ? :|

**Comment by alanpoulain** [comment]
@soyuka I think it's possible to use `JsonType` with a bad JSON for instance to throw a conversion exception.

**Comment by teohhanhui** [comment]
Or just use the `UuidType` with an invalid UUID. :laughing:

**Comment by soyuka** [comment]
thanks @Zowac

**Pull request merged**